groovy-language-server/src/test/java/com/palantir/ls/util/SourceWriterTest.java

```diff
@@ -0,0 +1,339 @@
+        assertEquals("my file contents\n", FileUtils.readFileToString(destination.toFile()));
+    }
+
+    @Test
+    public void testDidChanges_noChanges() throws IOException {
+        Path source = addFileToFolder(sourceFolder.getRoot(), "myfile.txt", "first line\nsecond line\n");
+        Path destination = destinationFolder.getRoot().toPath().resolve("myfile.txt");
+        SourceWriter writer = SourceWriter.of(source, destination);
+        List<TextDocumentContentChangeEvent> changes = Lists.newArrayList();
+        writer.applyChanges(changes);
+        assertEquals("first line\nsecond line\n", FileUtils.readFileToString(destination.toFile()));
+    }
+
+    @Test
+    public void testDidChanges_insertionBeginningOfLine() throws IOException {
+        Path source = addFileToFolder(sourceFolder.getRoot(), "myfile.txt", "first line\necond line\n");
+        Path destination = destinationFolder.getRoot().toPath().resolve("myfile.txt");
+        SourceWriter writer = SourceWriter.of(source, destination);
+        List<TextDocumentContentChangeEvent> changes = Lists.newArrayList();
+        changes.add(new TextDocumentContentChangeEventBuilder()
+                .range(Ranges.createRange(1, 0, 1, 0))
+                .rangeLength(1)
+                .text("s")
+                .build());
+        writer.applyChanges(changes);
+        assertEquals("first line\nsecond line\n", FileUtils.readFileToString(destination.toFile()));
+    }
+
+    @Test
+    public void testDidChanges_insertionEndOfLine() throws IOException {
+        Path source = addFileToFolder(sourceFolder.getRoot(), "myfile.txt", "first line\nsecond line\n");
+        Path destination = destinationFolder.getRoot().toPath().resolve("myfile.txt");
+        SourceWriter writer = SourceWriter.of(source, destination);
+        List<TextDocumentContentChangeEvent> changes = Lists.newArrayList();
+        changes.add(new TextDocumentContentChangeEventBuilder()
+                .range(Ranges.createRange(1, 20, 1, 20))
+                .rangeLength(13)
+                .text("small change\n")
+                .build());
+        writer.applyChanges(changes);
+        // Two new lines expected, one from the original contents and one from the change
+        assertEquals("first line\nsecond linesmall change\n\n", FileUtils.readFileToString(destination.toFile()));
+    }
+
+    @Test
+    public void testDidChanges_oneLineRange() throws IOException {
+        Path source = addFileToFolder(sourceFolder.getRoot(), "myfile.txt", "first line\nsecond line\n");
+        Path destination = destinationFolder.getRoot().toPath().resolve("myfile.txt");
+        SourceWriter writer = SourceWriter.of(source, destination);
+        List<TextDocumentContentChangeEvent> changes = Lists.newArrayList();
+        changes.add(new TextDocumentContentChangeEventBuilder()
+                .range(Ranges.createRange(0, 6, 0, 9))
```

Comment:
I guess based on our discussion, the range here should be (0,6,0,10) to get the expected results